src/parameter.cpp

```diff
@@ -36,50 +37,30 @@ bool Parameter::Init(const bool verbose) {
 Status Parameter::GetValues(ServerContext* context,
     ServerReaderWriter<Response, Request>* stream) {
 
-    char parhand_result[BUFSIZ];
+    GError *error=NULL;
+    AXParameter *ax_parameter = ax_parameter_new(APP_NAME, &error);
+    if (ax_parameter == NULL) {
+      ERRORLOG << "Error when creating axparameter: " << error->message << endl;
+      g_clear_error(&error);
+      return Status::CANCELLED;
+    }
 
     Request request;
     while (stream->Read(&request)) {
-      size_t pos = 0;
-      const char *parameter_value= NULL;
-      string parhand_cmd = "parhandclient get ";
-      string parameter_key = request.key().c_str();
-      const regex pattern("[a-zA-Z0-9.]+");
-      if(regex_match(parameter_key, pattern))
-      {
-        string parhandclient_cmd = parhand_cmd + parameter_key;
-
-        FILE *fp = popen(parhandclient_cmd.c_str(), "r"); 
-        if (!fp){
-          throw std::runtime_error("popen() failed!");
-        }
-        std::string value;
-        if ( fgets( parhand_result, BUFSIZ, fp ) != NULL ) {
-          value = parhand_result;
-          while ((pos = value.find('"', pos)) != std::string::npos)
-            value = value.erase(pos, 1);
-          parameter_value = value.c_str();
-        }
-        if (parameter_value != nullptr){
-          TRACELOG << request.key().c_str() << ": " << parameter_value << endl;
-        }
-        else {
-          parameter_value = "";
-          TRACELOG << request.key().c_str() << ": " << parameter_value << endl;
-        }
-
-        Response response;
-        response.set_value(parameter_value);
-        stream->Write(response);
-        pclose(fp);
-      }
-      else
-      {
-        TRACELOG << "No valid input request" << endl;
-        return Status(StatusCode::INVALID_ARGUMENT, "No valid input request");
+      char *parameter_value = NULL;
+      if (!ax_parameter_get(ax_parameter, request.key().c_str(),&parameter_value, &error)){
+        parameter_value = g_strdup("");
       }
+      TRACELOG << request.key().c_str() << ": " << parameter_value << endl;
+      Response response;
+      response.set_value(parameter_value);
+      stream->Write(response);
+      free(parameter_value);
+
     }
+    ax_parameter_free(ax_parameter);
+    g_clear_error(&error);
```

Comment:
If there are errors more than once from ax_parameter_get, this will only deallocate the last one.

ax_parameter_get() will only set an error when it returns false, so put g_clear_error() inside the if-block instead. No need for it here.